### PR TITLE
[16.0][ADD] web_widget_progressbar_gradient

### PR DESCRIPTION
**Pull request by rousseldenis**
Rendering example:

image

### CODE REVIEW

**Review by legalsylvain — APPROVED** [review]
Very Nice !
Thanks !

File: web_widget_progressbar_color/static/src/xml/progressbar.xml [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+    <t
+        t-name="web_widget_progressbar_color.ProgressBarFieldColor"
+        t-inherit="web.ProgressBarField"
+        t-inherit-mode="primary"
+        owl="1"
+    >
+        <xpath expr="//div[@t-attf-class]" position="attributes">
+            <attribute
+                name="t-attf-class"
+                add="{{ (state.currentValue > 40 and state.currentValue &lt;= 70) ? 'o_progress_warning' : state.currentValue > 70 ?  'o_progress_danger' : ''}}"
```
Comment (jbaudoux) [inline web_widget_progressbar_color/static/src/xml/progressbar.xml]:
```suggestion
                add="{{ (state.currentValue > 40 * state.maxValue/100 and state.currentValue &lt;= 70 * state.maxValue/100) ? 'o_progress_warning' : state.currentValue > 70 * state.maxValue/100 ?  'o_progress_danger' : ''}}"
```

**Review by jbaudoux — APPROVED** [review]
Nice.

**Review by jbaudoux — COMMENTED** [review]
Why not making a gradient progressbar instead ? Then you don't have to configure the thresholds

**Comment by rousseldenis** [comment]
> Why not making a gradient progressbar instead ? Then you don't have to configure the thresholds

Could you develop ?

**Comment by jbaudoux** [comment]
> > Why not making a gradient progressbar instead ? Then you don't have to configure the thresholds
> 
> Could you develop ?

Something like applying always this class:
```
.o_progressbar .o_progress  o_progressbar_gradient {
  background: linear-gradient(to right, green, yellow, red)
}
```

**Comment by rousseldenis** [comment]
> > > Why not making a gradient progressbar instead ? Then you don't have to configure the thresholds
> > 
> > 
> > Could you develop ?
> 
> Something like applying always this class:
> 
> ```
> .o_progressbar .o_progress  o_progressbar_gradient {
>   background: linear-gradient(to right, green, yellow, red)
> }
> ```

The result:

image

This is not the expecting result

**Comment by rousseldenis** [comment]
@legalsylvain @pedrobaeza @jbaudoux 

I will change the imlpementaiton to this (need a little bit tweaking) as this is our needs:

image

**Comment by rousseldenis** [comment]
I think also changing the module name to `web_widget_progressbar_gradient` to reflect more the final implementation.

**Comment by legalsylvain** [comment]
Hi @rousseldenis : looks nice ! 

A "matter of taste" remarks : 
- Generally, red is "bad / problematic", yellow is not "very good", and green is "good / allowed". for traffic lights, for example.
- Generally also, 0% means an unfulfilled objective, and 100% means a successful target
So, my proposal is to have red -> Yellow -> green. (instead of green -> yellow -> green).

However, just a suggestion. If you think it's better like that, let's move on.

**Comment by rousseldenis** [comment]
> * So, my proposal is to have red -> Yellow -> green. (instead of green -> yellow -> green).

The current implementation is to reflect the "occupation" which means that the less is better, the more is worse.

I'll try to add an option to the widget to allow the `inverse`mode (in fact, mine :sweat_smile: )

**Comment by rousseldenis** [comment]
@legalsylvain 

Now:

image


So both options are supported. The normal rendering is red to green. Inverse mode, green to red.

**Comment by legalsylvain** [comment]
Thanks ! 

It looks like there is a little offset when 100 is displayed.

**Comment by rousseldenis** [comment]
> It looks like there is a little offset when 100 is displayed.

That's also the case for the standard progressbar widget :sweat_smile:

**Comment by legalsylvain** [comment]
> That's also the case for the standard progressbar widget 😅

;-)

I think we can merge it. Agree ? Improvment can be done in other PRs.

**Comment by rousseldenis** [comment]
> Agree ?

Yes. Thanks !

File: web_widget_progressbar_gradient/static/src/scss/progressbar.scss [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    mask: linear-gradient(#fff 0 0);
+}
+
+.o_progressbar .o_progress .o_progressbar_gradient.o_inverse::before {
+    content: "";
+    position: absolute;
+    top: 0;
+    left: 0;
+    right: 0;
+    bottom: 0;
```
Comment (jbaudoux) [inline web_widget_progressbar_gradient/static/src/scss/progressbar.scss]: I think you can drop this. Maybe putting this section after the other would improve readability

Comment (rousseldenis) [reply web_widget_progressbar_gradient/static/src/scss/progressbar.scss]: What do you mean by `after the other` ?

**Review by apineux — APPROVED** [review]

File: web_widget_progressbar_gradient/static/src/scss/progressbar.scss (hunk 2 above)
Comment (jbaudoux) [reply web_widget_progressbar_gradient/static/src/scss/progressbar.scss]: the other css block just below that repeats the same

File: web_widget_progressbar_gradient/static/src/scss/progressbar.scss [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+.o_progressbar .o_progress .o_progressbar_gradient::before {
+    content: "";
+    position: absolute;
+    top: 0;
+    left: 0;
+    right: 0;
+    bottom: 0;
+    background-image: linear-gradient(to right, #dc3545, #ffc107, #198754);
+}
```
Comment (jbaudoux) [inline web_widget_progressbar_gradient/static/src/scss/progressbar.scss]: Like this:
```suggestion
.o_progressbar .o_progress .o_progressbar_gradient::before {
    content: "";
    position: absolute;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
    background-image: linear-gradient(to right, #dc3545, #ffc107, #198754);
}

.o_progressbar .o_progress .o_progressbar_gradient.o_inverse::before {
    background-image: linear-gradient(to right, #198754, #ffc107, #dc3545);
}
```

Comment (rousseldenis) [reply web_widget_progressbar_gradient/static/src/scss/progressbar.scss]: I will try but I had problems applying the inverse options in that case

**Review by FrancoMaxime — APPROVED** [review]
LGTM: code review

File: web_widget_progressbar_gradient/static/src/scss/progressbar.scss (hunk 3 above)
Comment (rousseldenis) [reply web_widget_progressbar_gradient/static/src/scss/progressbar.scss]: @jbaudoux Thanks. Working well.

**Pull request merged**